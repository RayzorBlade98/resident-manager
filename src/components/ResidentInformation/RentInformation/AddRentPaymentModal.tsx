import {
  Button, Grid, InputAdornment, TextField,
} from '@mui/material';
import React, { useState } from 'react';
import GenericModal from '../../GenericComponents/GenericModal/GenericModal';
import { DateInputString, DateString, DateUtils } from '_/types/date';
import { CurrencyInCents, convertCurrencyEurosToCents } from '_/utils/currency';
import {
  ValidationError,
  ValidationErrorMessages,
  createValidationFunction,
} from '_/utils/validation';

export interface AddRentPaymentModalProps {
  show: boolean;

  /**
   * Callback when the modal is closed
   */
  onClose: () => void;

  /**
   * Callback when the payment input gets submitted
   * @param paymentAmount Amount that was paid
   * @param paymentDate Date on which the payment was received
   */
  onSave: (paymentAmount: CurrencyInCents, paymentDate: DateString) => void;
}

interface RentPaymentInput {
  paymentAmount: CurrencyInCents | null;
  paymentDate: DateString;
}

/**
 * Modal that provides functionality to insert payment information
 */
function AddRentPaymentModal(props: AddRentPaymentModalProps): JSX.Element {
  const [payment, setPayment] = useState<RentPaymentInput>({
    paymentAmount: null,
    paymentDate: DateUtils.getCurrentDate(),
  });
  const [errors, setErrors] = useState<
  ValidationErrorMessages<RentPaymentInput>
  >({});

  const validatePayment = createValidationFunction<RentPaymentInput>({
    paymentAmount: [ValidationError.Null, ValidationError.LessEqualZero],
  });

  function paymentUpdater(field: string) {
    function updatePayment(event: React.ChangeEvent<HTMLInputElement>): void {
      let value: number | string = event.target.value;
      if (field === 'paymentAmount') {
        value = Number(value);
        value = convertCurrencyEurosToCents(value);
      } else if (field === 'paymentDate') {
        value = DateUtils.convertDateInputString(value as DateInputString);
      }
      const newPayment = {
        ...payment,
        [field]: value,
      };
      setPayment(newPayment);
      setErrors({
        ...errors,
        [field]: validatePayment(
          newPayment,
          field as keyof ValidationErrorMessages<RentPaymentInput>,
        ),
      });
    }
    return updatePayment;
  }

  function onSave(): void {
    const newErrors = validatePayment(
      payment,
    ) as ValidationErrorMessages<RentPaymentInput>;
    const numberOfErrors = Object.keys(newErrors).filter(
      (k) => !!newErrors[k as keyof ValidationErrorMessages<RentPaymentInput>],
    ).length;
    if (numberOfErrors > 0) {
      setErrors(newErrors);
      return;
    }
    props.onSave(payment.paymentAmount as CurrencyInCents, payment.paymentDate);
  }

  return (
    <GenericModal
      title="Neue Zahlung"
      show={props.show}
      onClose={props.onClose}
      size="sm"
    >
      {/* Body */}
      <form>
        <Grid container columnSpacing={2} rowSpacing={2}>
          <Grid item xs={12}>
            <TextField
              id="paymentAmount"
              label="Zahlungssumme"
              type="number"
              required
              onChange={paymentUpdater('paymentAmount')}
              InputProps={{
                endAdornment: <InputAdornment position="end">€</InputAdornment>,
              }}
              error={!!errors.paymentAmount}
              helperText={errors.paymentAmount || ''}
            />
          </Grid>
          <Grid item xs={12}>
            <TextField
              id="paymentDate"
              label="Zahlungsempfang"
              type="date"
              required
              defaultValue={DateUtils.convertToDateInputString(
                payment.paymentDate,
              )}
              onChange={paymentUpdater('paymentDate')}
              InputLabelProps={{
                shrink: true,
              }}
              error={!!errors.paymentDate}
              helperText={errors.paymentDate || ''}
            />
          </Grid>
        </Grid>
      </form>
      {/* Footer */}
      <Button variant="contained" onClick={() => onSave()}>
        Hinzufügen
      </Button>
    </GenericModal>
  );
}

export default AddRentPaymentModal;
