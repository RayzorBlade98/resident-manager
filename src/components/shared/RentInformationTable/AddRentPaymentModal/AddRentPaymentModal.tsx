import { Grid } from '@mui/material';
import React from 'react';
import RentInformationUtils from '../../../../utils/rent/rent.utils';
import { ValidationConstraint } from '../../../../utils/validation/constraints';
import Validator from '../../../../utils/validation/validator';
import CurrencyInputField from '_/components/form/CurrencyInputField/CurrencyInputField';
import FileSelect from '_/components/form/FileSelect/FileSelect';
import StandardDateField from '_/components/form/StandardDateField/StandardDateField';
import GenericModal from '_/components/generic/GenericModal/GenericModal';
import useFormValidation from '_/hooks/useFormValidation/useFormValidation';
import useResident from '_/hooks/useResident/useResident';
import { DocumentType } from '_/models/resident/document';
import { RentInformation } from '_/models/resident/rent';
import { Resident } from '_/models/resident/resident';
import { CurrencyInCents } from '_/utils/currency/currency.utils';

type AddRentPaymentModalProps = {
  /**
   * Whether to show the modal
   */
  show: boolean;

  /**
   * Callback when the modal is closed
   */
  onCloseModal: () => void;

  /**
   * Resident for which the rent payment should be added
   */
  resident: Resident;

  /**
   * Rent information for which the rent should be added
   */
  rentInformation: RentInformation;
};

/**
 * All values that can be submitted in the form
 */
interface AddRentPaymentInput {
  /**
   * Amount of rent that was paid
   */
  paymentAmount: CurrencyInCents;

  /**
   * Date of the payment
   */
  paymentDate: Date;

  /**
   * Path to the bank transfer document
   */
  bankTransferFile: string;
}

/**
 * Modal that provides functionality to insert payment information
 */
function AddRentPaymentModal(props: AddRentPaymentModalProps): JSX.Element {
  const { addRentPayment, addDocument } = useResident(props.resident.id);

  const {
    formInput,
    formErrors,
    formInputSetters,
    resetFormInput,
    FormSubmitButton,
  } = useFormValidation<AddRentPaymentInput>({
    formValidator: new Validator<AddRentPaymentInput>({
      paymentAmount: ValidationConstraint.Currency,
      paymentDate: ValidationConstraint.Defined,
      bankTransferFile: ValidationConstraint.DefinedFile,
    }),
    defaultFormInput: {
      paymentAmount: RentInformationUtils.getAmountToPay(props.rentInformation),
      paymentDate: new Date().toUTC(),
      bankTransferFile: undefined,
    },
    onSubmitSuccess: (values) => {
      void window.ipcAPI.persistence
        .uploadDocument(values.bankTransferFile, {
          type: 'resident',
          residentId: props.resident.id,
        })
        .then((bankTransferDocumentId) => {
          addDocument({
            name: `Mietzahlung ${props.rentInformation.dueDate.toString()}`,
            type: DocumentType.BankTransfer,
            creationDate: values.paymentDate,
            subjectDate: props.rentInformation.dueDate,
            id: bankTransferDocumentId,
          });
          addRentPayment({
            dueDate: props.rentInformation.dueDate,
            paymentAmount: values.paymentAmount,
            paymentDate: values.paymentDate,
            bankTransferDocumentId,
          });
          props.onCloseModal();
        });
    },
    submitButtonLabel: 'Hinzufügen',
  });

  return (
    <GenericModal
      title="Neue Zahlung"
      show={props.show}
      onClose={() => {
        props.onCloseModal();
        resetFormInput();
      }}
    >
      {/* Body */}
      <Grid container columnSpacing={2} rowSpacing={2}>
        <Grid item xs={6}>
          <CurrencyInputField
            required
            id="paymentAmount"
            label="Zahlungssumme"
            value={formInput.paymentAmount}
            onChange={formInputSetters.paymentAmount}
            errorMessage={formErrors.paymentAmount}
          />
        </Grid>
        <Grid item xs={6}>
          <StandardDateField
            required
            id="paymentDate"
            label="Zahlungsempfang"
            value={formInput.paymentDate}
            onChange={formInputSetters.paymentDate}
            errorMessage={formErrors.paymentDate}
          />
        </Grid>
        <Grid item xs={12}>
          <FileSelect
            id="bankTransferFile"
            label="Überweisung"
            value={formInput.bankTransferFile}
            onChange={formInputSetters.bankTransferFile}
            errorMessage={formErrors.bankTransferFile}
          />
        </Grid>
      </Grid>

      {/* Footer */}
      <FormSubmitButton />
    </GenericModal>
  );
}

export default AddRentPaymentModal;
