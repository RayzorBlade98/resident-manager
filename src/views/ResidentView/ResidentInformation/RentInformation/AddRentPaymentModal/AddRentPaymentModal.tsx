import React from 'react';
import { useRecoilValue, useResetRecoilState } from 'recoil';
// eslint-disable-next-line max-len
import GenericModal from '../../../../../components/GenericComponents/GenericModal/GenericModal';
import addRentPaymentState from '../states/add_rent_payment_state';
import AddRentPaymentButton from './AddRentPaymentButton/AddRentPaymentButton';
import AddRentPaymentForm from './AddRentPaymentForm/AddRentPaymentForm';

/**
 * Modal that provides functionality to insert payment information
 */
function AddRentPaymentModal(): JSX.Element {
  const rentPaymentState = useRecoilValue(addRentPaymentState);
  const resetRentPaymentState = useResetRecoilState(addRentPaymentState);

  return (
    <GenericModal
      title="Neue Zahlung"
      show={rentPaymentState.showModal}
      onClose={resetRentPaymentState}
      size="sm"
    >
      {/* Body */}
      <AddRentPaymentForm />
      {/* Footer */}
      <AddRentPaymentButton />
    </GenericModal>
  );
}

export default AddRentPaymentModal;
