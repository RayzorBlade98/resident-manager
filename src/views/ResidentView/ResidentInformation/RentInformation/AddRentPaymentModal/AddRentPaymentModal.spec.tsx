import { act, render } from '@testing-library/react';
import { generateImage } from 'jsdom-screenshot';
import React from 'react';
import { setRecoil } from 'recoil-nexus';
import addRentPaymentState from '../states/add_rent_payment_state';
import AddRentPaymentModal from './AddRentPaymentModal';
import ReactTestWrapper from '_/test/ReactTestWrapper';

describe('AddRentPaymentModal', () => {
  test('should match snapshot', async () => {
    // Arrange
    render(
      <ReactTestWrapper>
        <AddRentPaymentModal />
      </ReactTestWrapper>,
    );
    act(() => {
      setRecoil(addRentPaymentState, (state) => ({
        ...state,
        showModal: true,
        formValidation: {
          ...state.formValidation,
          formInput: {
            ...state.formValidation.formInput,
            paymentDate: new Date(2023, 5, 11),
          },
        },
      }));
    });

    // Assert
    expect(
      await generateImage({ viewport: { width: 650, height: 400 } }),
    ).toMatchImageSnapshot();
  });
});
