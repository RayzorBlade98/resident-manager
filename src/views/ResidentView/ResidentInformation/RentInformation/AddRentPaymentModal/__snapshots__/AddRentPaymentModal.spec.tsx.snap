// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddRentPaymentModal should match snapshot 1`] = `
<div>
  <div
    class="modal fade show"
    style="display: block; pointer-events: initial;"
    tabindex="-1"
  >
    <div
      class="modal-dialog modal-sm"
    >
      <div
        class="modal-content"
      >
        <div
          class="modal-header"
        >
          <h5
            class="modal-title"
          >
            Neue Zahlung
          </h5>
          <button
            class="ripple ripple-surface btn-close"
            role="button"
          />
        </div>
        <div
          class="modal-body"
        >
          <form>
            <div
              class="MuiGrid-root MuiGrid-container css-mhc70k-MuiGrid-root"
            >
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="false"
                    for="paymentAmount"
                    id="paymentAmount-label"
                  >
                    Zahlungssumme
                  </label>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl MuiInputBase-adornedEnd css-o9k5xi-MuiInputBase-root-MuiOutlinedInput-root"
                  >
                    <input
                      aria-invalid="false"
                      class="MuiInputBase-input MuiOutlinedInput-input MuiInputBase-inputAdornedEnd css-nxo287-MuiInputBase-input-MuiOutlinedInput-input"
                      id="paymentAmount"
                      type="number"
                      value=""
                    />
                    <div
                      class="MuiInputAdornment-root MuiInputAdornment-positionEnd MuiInputAdornment-outlined MuiInputAdornment-sizeMedium css-1laqsz7-MuiInputAdornment-root"
                    >
                      <p
                        class="MuiTypography-root MuiTypography-body1 css-1pnmrwp-MuiTypography-root"
                      >
                        €
                      </p>
                    </div>
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="css-yjsfm1"
                      >
                        <span>
                          Zahlungssumme
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
              <div
                class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 css-1idn90j-MuiGrid-root"
              >
                <div
                  class="MuiFormControl-root MuiTextField-root css-1u3bzj6-MuiFormControl-root-MuiTextField-root"
                >
                  <label
                    class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled Mui-required MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-outlined css-1sumxir-MuiFormLabel-root-MuiInputLabel-root"
                    data-shrink="true"
                    for="paymentDate"
                    id="paymentDate-label"
                  >
                    Zahlungsempfang
                    <span
                      aria-hidden="true"
                      class="MuiFormLabel-asterisk MuiInputLabel-asterisk css-wgai2y-MuiFormLabel-asterisk"
                    >
                       
                      *
                    </span>
                  </label>
                  <div
                    class="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-formControl css-9ddj71-MuiInputBase-root-MuiOutlinedInput-root"
                  >
                    <input
                      aria-invalid="false"
                      autocomplete="off"
                      class="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
                      id="paymentDate"
                      inputmode="text"
                      placeholder="TT.MM.JJJJ"
                      required=""
                      type="text"
                      value="28.05.2023"
                    />
                    <fieldset
                      aria-hidden="true"
                      class="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
                    >
                      <legend
                        class="css-14lo706"
                      >
                        <span>
                          Zahlungsempfang
                           
                          *
                        </span>
                      </legend>
                    </fieldset>
                  </div>
                </div>
              </div>
            </div>
          </form>
        </div>
        <div
          class="modal-footer"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium Mui-focusVisible MuiButton-root MuiButton-contained MuiButton-containedError MuiButton-sizeMedium MuiButton-containedSizeMedium css-1sszw6i-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Hinzufügen
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            >
              <span
                class="css-y4cjyz-MuiTouchRipple-ripple MuiTouchRipple-ripple MuiTouchRipple-rippleVisible MuiTouchRipple-ripplePulsate"
                style="width: 1px; height: 1px; top: -0.5px; left: -0.5px;"
              >
                <span
                  class="MuiTouchRipple-child MuiTouchRipple-childPulsate"
                />
              </span>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
</div>
`;
